doc/parameter_server.md

```diff
@@ -0,0 +1,229 @@
+- Master defines a hash strategy from parameter key to pserver id 
+
+Workflow:
+
+
+1. Master creates several pservers and workers according to user's configuration
+2. Master generates parameter sharding strategy based on model definition provided by user
+3. Master initializes parameter, and send to pservers
+4. Master triggers pservers and workers and start training
+5. Master monitors cluster, and increase or decrease workerers according to priority
+
+
+## Pserver
+
+- Pserver provides kv store service, and workers could push/pull <key, value> to the pserver
+- Pserver provides optimize service, apply gradients to parameters and update back to kv store
+
+
+Workflow:
+
+1. Pserver waits for workers pushing <key, gradient> to pserver
+2. Pserver gets the key and queries kv store to get corresponding parameter
+3. Pserver calls optimizer, applys <key, gradient> to <key, parameter>
+4. Pserver updates <key, parameter> back to kv store
+5. Pserver saves model checkpoint to disk periodially
+
+
+**Note**
+
+We could implement the kv store by ourselves, or we could use some already solution, such as redis.
+
+## Worker
+
+- Worker defines a forward/backward computation proccess
+- Worker defines a dataloader module
+
+Workflow:
+
+1. Before forward layer starting, worker sends key to pserver, and pull <key, parameter> back from pserver
+2. Worker starts forward layer computation
+3. After backward layer finishing, worker generates <key, gradient>
+4. Worker pushes <key, gradient> to pserver
+
+
+There is also another kind of worker who does an evalution job. It define its own evalution computation process and dataloader of validation data. It pulls <key, parameter> from pserver, but never push data to pserver.
+
+
+**Note**
+
+There could be some same item id in a minibatch data. So some gradient vector of embedding table will have the same item id. We need to sum these gradient before pushing to pserver.
+
+
+## Delayed model updating
+
+In order to reduce the communication overhead between workers and pservers, we propose a stragety called delayed model updating.
+
+A worker runs a round of forward/backward computation using its local model, and keep the gradients in its local. After several rounds finishing, it push gradients to pservers.
+
+**Note**
+
+Since local model is only a part of global model, in some circumstance, worker still has to pull embedding vector parameter from pservers if there exits unknow item ids in a minibatch data.
+
+In async mode, this will lead to relative newer embedding part parameters, but relative older other part parameters.
+
+
+## Elastic scheduling
+
+### Analysis
+
+First, we analyse the hardware resources that pserver and worker consume.
+
+
+| role| CPU | memory | network bandwidth |  disk |
+| :----  |:----  |:---|:---|:---|
+| worker  | high| medium |  medium | low|
+| pserver | low | medium | high| medium |
+
```

Comment:
Different models may result in different hw resource requirements.
Worker memory depends on the model and the minibatch size, may be very high. For example, MaskRCNN model can only support `minibatch=2` for a GPU with 16G memory in TensorFlow with default setting.

Reply:
Exactly, here we are focusing on recommendation scenario, which big embedding table is stored at pserver, and worker only has a partial model.  So the memory is `relative medium`.